File: 프로그래머스/BFSDFS/타겟넘버.py
# 프로그래머스 타겟 넘버
# 21.01.19

def solution(numbers, target):
    answer = DFS(numbers, target, 0, 0)
    return answer


def DFS(numbers, target, idx, sum):
    if idx == len(numbers):
        if sum == target:
            return 1
        else:
            return 0
    else:
        return DFS(numbers, target, idx + 1, sum+numbers[idx]) + DFS(numbers, target, idx + 1, sum-numbers[idx])

File: 프로그래머스/BFSDFS/타겟넘버.cpp
/*----------------------
|프로그래머스 타겟 넘버
|21.01.26
----------------------*/

#include <string>
#include <vector>

using namespace std;

int DFS(int idx, vector<int> &num, int sum, int &target)
{
    if (idx == num.size())
    {
        if (sum == target) return 1;
        else return 0;
    }

    return DFS(idx+1, num, sum+num[idx], target) + DFS(idx+1, num, sum-num[idx], target);
}
int solution(vector<int> numbers, int target) 
{
    int answer = DFS(0, numbers, 0, target);

    return answer;
}
